pkg/webhooks/validation/server/server.go

```diff
@@ -0,0 +1,500 @@
+	keyCertWatcher *fsnotify.Watcher
+
+	mu   sync.RWMutex
+	cert *tls.Certificate
+
+	// pilot
+	descriptor   schema.Set
+	domainSuffix string
+
+	// mixer
+	validator store.BackendValidator
+
+	server   *http.Server
+	keyFile  string
+	certFile string
+}
+
+// Reload the server's cert/key for TLS from file and save it for later use by the https server.
+func (wh *Webhook) reloadKeyCert() {
+	pair, err := reloadKeyCert(wh.certFile, wh.keyFile)
+	if err != nil {
+		return
+	}
+
+	wh.mu.Lock()
+	wh.cert = pair
+	wh.mu.Unlock()
+}
+
+// Reload the server's cert/key for TLS from file.
+func reloadKeyCert(certFile, keyFile string) (*tls.Certificate, error) {
+	pair, err := tls.LoadX509KeyPair(certFile, keyFile)
+	if err != nil {
+		reportValidationCertKeyUpdateError(err)
+		scope.Warnf("Cert/Key reload error: %v", err)
+		return nil, err
+	}
+
+	reportValidationCertKeyUpdate()
+	scope.Info("Cert and Key reloaded")
+
+	var row int
+	for _, cert := range pair.Certificate {
+		if x509Cert, err := x509.ParseCertificates(cert); err != nil {
+			scope.Infof("x509 cert [%v] - ParseCertificates() error: %v\n", row, err)
+			row++
+		} else {
+			for _, c := range x509Cert {
+				scope.Infof("x509 cert [%v] - Issuer: %q, Subject: %q, SN: %x, NotBefore: %q, NotAfter: %q\n",
+					row, c.Issuer, c.Subject, c.SerialNumber,
+					c.NotBefore.Format(time.RFC3339), c.NotAfter.Format(time.RFC3339))
+				row++
+			}
+		}
+	}
+	return &pair, nil
+}
+
+// New creates a new instance of the admission webhook server.
+func New(p Options) (*Webhook, error) {
+	if p.Mux != nil {
+		wh := &Webhook{
+			descriptor: p.PilotDescriptor,
+			validator:  p.MixerValidator,
+		}
+
+		p.Mux.HandleFunc("/admitpilot", wh.serveAdmitPilot)
+		p.Mux.HandleFunc("/admitmixer", wh.serveAdmitMixer)
+		return wh, nil
+	}
+	pair, err := reloadKeyCert(p.CertFile, p.KeyFile)
+	if err != nil {
+		return nil, err
+	}
+
+	// Configuration must be updated whenever the caBundle changes. Watch the parent directory of
+	// the target files so we can catch symlink updates of k8s secrets.
+	keyCertWatcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	for _, file := range []string{p.CertFile, p.KeyFile} {
+		watchDir, _ := filepath.Split(file)
+		if err := keyCertWatcher.Watch(watchDir); err != nil {
+			return nil, fmt.Errorf("could not watch %v: %v", file, err)
+		}
+	}
+
+	wh := &Webhook{
+		server: &http.Server{
+			Addr: fmt.Sprintf(":%v", p.Port),
+		},
+		keyFile:        p.KeyFile,
+		certFile:       p.CertFile,
+		keyCertWatcher: keyCertWatcher,
+		cert:           pair,
+		descriptor:     p.PilotDescriptor,
+		validator:      p.MixerValidator,
+	}
+
+	// mtls disabled because apiserver webhook cert usage is still TBD.
+	wh.server.TLSConfig = &tls.Config{GetCertificate: wh.getCert}
+	h := http.NewServeMux()
+	h.HandleFunc("/admitpilot", wh.serveAdmitPilot)
+	h.HandleFunc("/admitmixer", wh.serveAdmitMixer)
+	h.HandleFunc(HTTPSHandlerReadyPath, wh.serveReady)
```

Comment:
this might conflict with other components? Not sure what would happen if that is true, or if it is even true. Longer term we probably need a top level /ready endpoint that aggregates readiness of all components

Reply:
Nevermind, this is only when run on its own mux

Reply:
It shouldn't conflict. Longer term we do need to properly aggregate readiness.